### PR TITLE
Add `drop_last=True` in DataLoaders inside `ls test_train*`

### PR DESCRIPTION
**Pull request by taylanbil**
without drop_last, loaders produce a batch of different size in last step.

### CODE REVIEW

**Comment by taylanbil** [comment]
When investigating [DistributedSampler's `ceil`](https://github.com/pytorch/pytorch/blob/06bb74ce96f80856ded884282c17777f61fd3212/torch/utils/data/distributed.py#L39), I ran the following test:


>So in test_train_mp_mnist.py, if I add drop_last=True to https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L87-L92, then the last batch is indeed dropped. There are 58 batches w/ drop_last, as opposed to 59 without. Similarly with 1 core, drop_last=True will make 468 batches and without that we have 469 (with different shape btw)
```
ord 0 batch 467 torch.Size([128, 1, 28, 28])
ord 0 batch 468 torch.Size([96, 1, 28, 28])
```
> AFAIU, drop_last in DataLoader needs DistributedSampler to round up. That ensures same length in all processes, and then it can safely drop last. WDYT?

**Comment by dlibenzi** [comment]
> When investigating [DistributedSampler's `ceil`](https://github.com/pytorch/pytorch/blob/06bb74ce96f80856ded884282c17777f61fd3212/torch/utils/data/distributed.py#L39), I ran the following test:
> 
> > So in test_train_mp_mnist.py, if I add drop_last=True to [/test/test_train_mp_mnist.py@`master`#L87-L92](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L87-L92) [](/pytorch/xla/blob/HEAD@{2019-10-31T19:03:09Z}/test/test_train_mp_mnist.py#L87-L92), then the last batch is indeed dropped. There are 58 batches w/ drop_last, as opposed to 59 without. Similarly with 1 core, drop_last=True will make 468 batches and without that we have 469 (with different shape btw)
> 
> ```
> ord 0 batch 467 torch.Size([128, 1, 28, 28])
> ord 0 batch 468 torch.Size([96, 1, 28, 28])
> ```
> 
> > AFAIU, drop_last in DataLoader needs DistributedSampler to round up. That ensures same length in all processes, and then it can safely drop last. WDYT?

The only bad condition is if only M<N of the N cores end up firing the last computation.
The fact that they have different shapes should not matter.
Even though different shapes, the gradient must be of the same shape, since weights are.
So we should be safe from an AllReduce POV.

Do we ever get into firing M<N computations?

**Comment by taylanbil** [comment]
> So we should be safe from an AllReduce POV.

Yeah this PR is not a bug fix, just an "optimization", reducing compiles. We don't ever get into the M<N situation, no. I can drop the PR if needed.

**Comment by dlibenzi** [comment]
> > So we should be safe from an AllReduce POV.
> 
> Yeah this PR is not a bug fix, just an "optimization", reducing compiles. We don't ever get into the M<N situation, no. I can drop the PR if needed.

I'd feel more comfortable if this was an argument into the script, instead of being bolted in.

**Comment by taylanbil** [comment]
makes sense, I'll adjust accordingly.

**Comment by taylanbil** [comment]
the ci failure is unrelated to this pr.
